src/lib/chawathe98_matcher.rs

```diff
@@ -46,35 +46,2262 @@
+        */
+
+        return (edges_still_left, edges_to_be_pruned);
+    }
+}
+
+/// Creates a simple tree for tests which is FromNodeId
+pub fn create_arena() -> Arena<String, FromNodeId> {
+    let mut arena = Arena::new();
+    let root = arena.new_node(
+        String::from("Expr"),
+        String::from("+"),
+        None,
+        None,
+        None,
+        None,
+    );
+
+    let n2 = arena.new_node(
+        String::from("Expr"),
+        String::from("*"),
+        None,
+        None,
+        None,
+        None,
+    );
+    n2.make_child_of(root, &mut arena).unwrap();
+    let n3 = arena.new_node(
+        String::from("INT"),
+        String::from("3"),
+        None,
+        None,
+        None,
+        None,
+    );
+    n3.make_child_of(n2, &mut arena).unwrap();
+    let n4 = arena.new_node(
+        String::from("INT"),
+        String::from("4"),
+        None,
+        None,
+        None,
+        None,
+    );
+    n4.make_child_of(n2, &mut arena).unwrap();
+
+    let n1 = arena.new_node(
+        String::from("INT"),
+        String::from("1"),
+        None,
+        None,
+        None,
+        None,
+    );
+    n1.make_child_of(root, &mut arena).unwrap();
+    arena
+}
+
+/// Creates a simple tree for tests which is ToNodeId
+pub fn create_arena_2() -> Arena<String, ToNodeId> {
+    let mut arena = Arena::new();
+    let root = arena.new_node(
+        String::from("Expr"),
+        String::from("+"),
+        None,
+        None,
+        None,
+        None,
+    );
+
+    let n2 = arena.new_node(
+        String::from("Expr"),
+        String::from("*"),
+        None,
+        None,
+        None,
+        None,
+    );
+    n2.make_child_of(root, &mut arena).unwrap();
+    let n3 = arena.new_node(
+        String::from("INT"),
+        String::from("3"),
+        None,
+        None,
+        None,
+        None,
+    );
+    n3.make_child_of(n2, &mut arena).unwrap();
+    let n4 = arena.new_node(
+        String::from("INT"),
+        String::from("4"),
+        None,
+        None,
+        None,
+        None,
+    );
+    n4.make_child_of(n2, &mut arena).unwrap();
+
+    let n1 = arena.new_node(
+        String::from("INT"),
+        String::from("1"),
+        None,
+        None,
+        None,
+        None,
+    );
+    n1.make_child_of(root, &mut arena).unwrap();
+
+    arena
+}
+
+/// Test that two nodes have the same label and type.
+fn eq<T: Clone + Debug + Eq>(
+    n1: &NodeId<FromNodeId>,
+    arena1: &Arena<T, FromNodeId>,
+    n2: &NodeId<ToNodeId>,
+    arena2: &Arena<T, ToNodeId>,
+) -> bool {
+    arena1[*n1].label == arena2[*n2].label && arena1[*n1].ty == arena2[*n2].ty
+}
+
+/// Test that two nodes have the same label.
+fn eq_label<T: Clone + Debug + Eq>(
+    n1: &NodeId<FromNodeId>,
+    arena1: &Arena<T, FromNodeId>,
+    n2: &NodeId<ToNodeId>,
+    arena2: &Arena<T, ToNodeId>,
+) -> bool {
+    arena1[*n1].ty == arena2[*n2].ty
```

Comment:
This tests the types of the nodes, not their labels! So either the code here is wrong or the function name and comment!